optaplanner-core/src/main/java/org/optaplanner/core/impl/util/IntervalSplitPoint.java

```diff
@@ -0,0 +1,131 @@
+    TreeSet<Interval<_IntervalValue, _PointValue>> intervalsEndingAtSplitPointSet;
+
+    public IntervalSplitPoint(_PointValue splitPoint) {
+        this.splitPoint = splitPoint;
+    }
+
+    protected void createCollections() {
+        startIntervalToCountMap = new IdentityHashMap<>();
+        endIntervalToCountMap = new IdentityHashMap<>();
+        intervalsStartingAtSplitPointSet = new TreeSet<>(
+                Comparator.<Interval<_IntervalValue, _PointValue>, _PointValue> comparing(Interval::getEnd)
+                        .thenComparingInt(interval -> System.identityHashCode(interval.value)));
+        intervalsEndingAtSplitPointSet = new TreeSet<>(
+                Comparator.<Interval<_IntervalValue, _PointValue>, _PointValue> comparing(Interval::getStart)
+                        .thenComparingInt(interval -> System.identityHashCode(interval.value)));
+    }
+
+    public boolean addIntervalStartingAtSplitPoint(Interval<_IntervalValue, _PointValue> interval) {
+        startIntervalToCountMap.merge(interval.value, 1, Integer::sum);
+        return intervalsStartingAtSplitPointSet.add(interval);
+    }
+
+    public void removeIntervalStartingAtSplitPoint(Interval<_IntervalValue, _PointValue> interval) {
+        Integer newCount = startIntervalToCountMap.computeIfPresent(interval.value, (key, count) -> {
+            if (count > 1) {
+                return count - 1;
+            }
+            return null;
+        });
+        if (null == newCount) {
+            intervalsStartingAtSplitPointSet.remove(interval);
+        }
+    }
+
+    public boolean addIntervalEndingAtSplitPoint(Interval<_IntervalValue, _PointValue> interval) {
+        endIntervalToCountMap.merge(interval.value, 1, Integer::sum);
+        return intervalsEndingAtSplitPointSet.add(interval);
+    }
+
+    public void removeIntervalEndingAtSplitPoint(Interval<_IntervalValue, _PointValue> interval) {
+        Integer newCount = endIntervalToCountMap.computeIfPresent(interval.value, (key, count) -> {
+            if (count > 1) {
+                return count - 1;
+            }
+            return null;
+        });
+        if (null == newCount) {
+            intervalsEndingAtSplitPointSet.remove(interval);
+        }
+    }
+
+    public boolean containsIntervalStarting(Interval<_IntervalValue, _PointValue> interval) {
+        return intervalsStartingAtSplitPointSet.contains(interval);
+    }
+
+    public boolean containsIntervalEnding(Interval<_IntervalValue, _PointValue> interval) {
+        return intervalsEndingAtSplitPointSet.contains(interval);
+    }
+
+    public Iterator<_IntervalValue> getValuesStartingFromSplitPointIterator() {
+        return intervalsStartingAtSplitPointSet.stream()
+                .flatMap(interval -> IntStream.range(0, startIntervalToCountMap.get(interval.value))
+                        .mapToObj((index) -> interval.value))
+                .iterator();
+    }
```

Comment:
The iterator here seems unnecessary - streams are already iterable.
I do not see where this is used, perhaps there is some reason for it, but unless it's necessary, I'd prefer removing the indirection.